### DISCUSSION TITLE
Perfomance optimization

### DISCUSSION THREAD

**Discussion by pahaz** [keystonejs/keystone#5890] [Questions] [answered]
Hi 👋 ! I have Keystone based project. At the moment I have a 400 - 800 ms response time for my queries.

It looks not so good and this does not allow us to scale fast.

The main reason for that is a large number of database queries. The keystone doesn't try to cache any `resolved` data during one query. 

Has anyone ever had to deal with the keystone performance optimization?

PS: Currently, we have about 1-2 HTTP requests per second which generate about 400 database requests per second.

PS: I know about cache: https://www.keystonejs.com/guides/cache-hints/
But I need maybe anyone have another experience or cases

**Comment by MurzNN** [comment]
That strange, on my setup I see that Keystone is very robust, I see 20-40 ms of GraphQL replies, and not so much SQL queries... For example, simple GraphQL query:
```graphql
query {
  allUsers {
    id, email
  }
}
```
provokes only one SQL query:
```
prisma:query SELECT "public"."User"."id", "public"."User"."displayName", "public"."User"."firstName", "public"."User"."middleName", "public"."User"."lastName", "public"."User"."email", "public"."User"."password", "public"."User"."createdAt", "public"."User"."updatedAt", "public"."User"."createdBy", "public"."User"."updatedBy" FROM "public"."User" WHERE 1=1 OFFSET $1
```

You can log all SQL queries via adding `enableLogging: true` to db config.

Can you provide more details, which kinds of queries does Keystone execute on single request? And which type of request - read (query) or write (mutation), how many relations did you touch in it? 

And which version of Keystone did you use - 5 or 6?

**Comment by MurzNN** [comment]
Also I filled suggestion about making hooks as sync https://github.com/keystonejs/keystone/discussions/5894 that must slightly increase performance too.

**Comment by MurzNN** [comment]
Also I've figured out problem with relationships, here is bug report: https://github.com/keystonejs/keystone/issues/6035

**Comment by raveling** [comment]
Thanks @MurzNN 
@pahaz assume your perf is OK as you haven't replied in some time.

**Comment by pahaz** [comment]
Hi, we have a model like so:

```
keystone.createList('User', {
       ....
})
keystone.createList('Ticket', {
        assignee: {
            schemaDoc: 'Assignee/responsible employee/user who must ensure that the issue is fulfilled',
            type: Relationship,
            ref: 'User',
        },
        executor: {
            schemaDoc: 'Executor employee/user who perform the issue',
            type: Relationship,
            ref: 'User',
        },
        ....
})
```

And we have a query like so:
```
allTickets(...) {
    assignee {
        ...
    }
    executor {
        ...
    }
}
```

This GQL query produces N+1 sub-queries for any Relation.

**Reply by molomby** [reply]
Hey pahaz, your code examples indicate you're on [Keystone v5](https://github.com/keystonejs/keystone-5) – this repo is exclusively for Keystone v6. These two KS version use a completely different query builder/data access layer (mongoose/knex in Keystone 5, [Prisma](https://www.prisma.io/) in Keystone 6) so the answers you're getting here won't carry across.

**Comment by bladey** [comment]
Hey @pahaz,

If this is still an issue could you please raise it in the Keystone Community Slack and we'll assist - https://community.keystonejs.com

Thank you!

**Comment by gwyneplaine** [comment, marked as answer]
> Hey pahaz, your code examples indicate you're on [Keystone v5](https://github.com/keystonejs/keystone-5) – this repo is exclusively for Keystone v6. These two KS version use a completely different query builder/data access layer (mongoose/knex in Keystone 5, [Prisma](https://www.prisma.io/) in Keystone 6) so the answers you're getting here won't carry across.
[Answered by @molomby](https://github.com/keystonejs/keystone/discussions/5890#discussioncomment-2427525)
